src/Microsoft.AspNetCore.SignalR/HubEndPoint.cs

```diff
@@ -59,24 +59,54 @@ public class HubEndPoint<THub, TClient> : IInvocationBinder where THub : Hub<TCl
 
         public async Task OnConnectedAsync(ConnectionContext connection)
         {
-            await ProcessNegotiate(connection);
+            var output = Channel.CreateUnbounded<byte[]>();
+            var connectionContext = new HubConnectionContext(output, connection);
+
+            await ProcessNegotiate(connectionContext);
```

Comment:
The problem is that negotiation happens over a connection that has already been opened. You can set mode in WebSocketOptions but it is before negotiation happens and this is the client who requests a protocol so you don't know what to set it to. For webSockets we could try always sending binary messages but it requires converting to string on the client. This won't work with SSE though. SSE either needs to reject binary protocols or base64 encode but since the transport is decoupled from connection neither side knows if it should do that.